### PR TITLE
Checkpoint V3.5.1

### PR DESCRIPTION
**Pull request by PanLeon**
## TRELLO

* [Main Trello ticket](https://trello.com/b/BJDdC443/mytrip)

## Code reviewers

- [ ] @Ashiato
- [ ] @black-art
- [ ] @hrytskivr
- [ ] @KarnaushenkoVO
- [ ] @PanLeon
- [ ] @spycheast
- [ ] @yuriiZherebukh

### Second Level Review

- [ ] @byshovets
- [ ] @lhalam

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=90%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: myTrip/checkpoint/views.py [hunk 1]
```diff
@@ -29,18 +29,21 @@ def get(self, request, trip_id, checkpoint_id=None):
         checkpoints_list = [check.to_dict() for check in checkpoints]
         return JsonResponse(checkpoints_list, status=200, safe=False)
 
+        # return JsonResponse(checkpoints_list, status=401, safe=False)
```
Comment (lhalam) [inline myTrip/checkpoint/views.py]: delete this!

File: myTrip/checkpoint/views.py [hunk 2]
```diff
@@ -50,6 +53,8 @@ def post(self, request, trip_id):
                                    trip=trip)
         return JsonResponse(result.to_dict(), status=200)
 
+        # return HttpResponse(status=403)
```
Comment (lhalam) [inline myTrip/checkpoint/views.py]: delete this!

File: myTrip/checkpoint/views.py [hunk 3]
```diff
@@ -68,9 +73,10 @@ def put(self, request, checkpoint_id, trip_id):
                           title=data.get('title'),
                           description=data.get('description'),
                           position_number=data.get('position_number'))
-
         return JsonResponse(checkpoint.to_dict(), status=200)
 
+        # return HttpResponse(status=405)
```
Comment (lhalam) [inline myTrip/checkpoint/views.py]: delete this!

File: myTrip/checkpoint/views.py [hunk 4]
```diff
@@ -85,3 +91,5 @@ def delete(self, request, checkpoint_id, trip_id):
         if not result:
             return HttpResponse(status=404)
         return HttpResponse(status=200)
+
+        # return HttpResponse(status=406)
```
Comment (lhalam) [inline myTrip/checkpoint/views.py]: delete this!

File: myTrip/static/src/checkpoint/actions/index.js [hunk 5]
```diff
@@ -1,17 +1,20 @@
...
-export function getAllCheckpoints(trip_id) {
+export const getAllCheckpoints = trip_id => {
     return {
         type: 'GET-CHECKPOINTS',
         payload: service.getAllCheckpoints(trip_id)
             .then(function(response){
                 return response.data
-            })
+            },function(error){
+            dispatch({type:'ERROR',
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/actions/index.js]: fix tab/space

File: myTrip/static/src/checkpoint/actions/index.js [hunk 6]
```diff
@@ -1,17 +1,20 @@
...
+                    payload:error.response})
+        })
     }
 };
 
-export const checkpointDetails = (checkpoint) => {
+export const checkpointDetails = checkpoint=> {
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/actions/index.js]: add space `checkpoint =>`

File: myTrip/static/src/checkpoint/actions/index.js [hunk 7]
```diff
@@ -25,10 +28,9 @@ export const closeDetails = () => {
     }
 };
 
-export const createCheckpointUpdateList =  (longitude,latitude,title,description,position_number,source_url, trip_id) =>{
-    return {
-        type: 'CREATE-CHECKPOINT-UPDATE_LIST',
-        payload: service.createCheckpoint(longitude,
+export const createCheckpointUpdateList = (longitude,latitude,title,description,position_number,source_url,trip_id)=>{
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/actions/index.js:30]: add spaces `) => {`

File: myTrip/static/src/checkpoint/actions/index.js [hunk 8]
```diff
@@ -37,17 +39,20 @@ export const createCheckpointUpdateList =  (longitude,latitude,title,description
...
+                    payload:response.data})
+            })
+        },function(error){
+            dispatch({type:'ERROR',
+                    payload:error.response})
         })
     }
 };
 
 export const updateCheckpointUpdateList =  (longitude,latitude,title,description,position_number,source_url,trip_id,checkpoint_id) =>{
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/actions/index.js]: add spaces ) => {

File: myTrip/static/src/checkpoint/actions/index.js [hunk 9]
```diff
@@ -57,9 +62,13 @@ export const updateCheckpointUpdateList =  (longitude,latitude,title,description
...
         .then(function(response){
             return service.getAllCheckpoints(trip_id)
-                .then(function(response){
-                    return response.data
-                });
+            .then(function(response){
+                dispatch({type:'UPDATE-CHECKPOINT-UPDATE_LIST',
+                    payload:response.data})
+            })
+        },function(error){
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/actions/index.js]: fix tab/space

File: myTrip/static/src/checkpoint/actions/index.js [hunk 10]
```diff
@@ -79,6 +88,9 @@ export const deleteUpadateList = (id, trip_id) =>{
                     payload:response.data,
                     details:status})
             })
+        },function(error){
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/actions/index.js]: fix tab/space

File: myTrip/static/src/checkpoint/addCheckpoint.js [hunk 11]
```diff
@@ -3,65 +3,95 @@ import {connect} from 'react-redux';
...
+
     }
 
     addPoint(){
-        const longitude = 23.999003;
-        const latitude = 49.832721;
-        const title = 'test4';
-        const description = 'test';
-        const position_number = 1;
+        console.log('addpoint',this.props.checkpoints)
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/addCheckpoint.js]: delete this!

File: myTrip/static/src/checkpoint/addCheckpoint.js [hunk 12]
```diff
@@ -3,65 +3,95 @@ import {connect} from 'react-redux';
...
         const source_url = '';
         const tripId = this.props.trip.id
 
-        navigator.geolocation.getCurrentPosition( 
-            data => {
-                this.props.createCheckpointUpdateList(
-                    data.coords.longitude,
-                    data.coords.latitude,
+        service.CurrentPosition.then(response =>{
+            console.log('addpoint gromise',response.lng)
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/addCheckpoint.js]: delete this!

File: myTrip/static/src/checkpoint/addCheckpoint.js [hunk 13]
```diff
@@ -3,65 +3,95 @@ import {connect} from 'react-redux';
...
                     description,
                     position_number,
                     source_url,
                     tripId
-                )
-            },
+                )},
             err => {
-                console.log(err.message);
+                console.log('addpoint gromise',err.lng)
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/addCheckpoint.js]: delete this!

File: myTrip/static/src/checkpoint/checkpoint.service.js [hunk 14]
```diff
@@ -31,9 +30,30 @@ function getAllCheckpoints(trip_id) {
     return axios.get(url+trip_id+'/checkpoint/');
 }
 
+var CurrentPosition = new Promise(function(resolve, reject) {
+    //default center
+        const center ={lat:49.832721,lng:23.999003}
+        navigator.geolocation.getCurrentPosition( 
+                        data => {
+                            console.log({lat:data.coords.latitude,lng:data.coords.longitude})
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/checkpoint.service.js]: delete this!

File: myTrip/static/src/checkpoint/checkpoint.service.js [hunk 15]
```diff
@@ -31,9 +30,30 @@ function getAllCheckpoints(trip_id) {
...
+                            center.lat = data.coords.latitude;
+                            center.lng = data.coords.longitude
+                            resolve(center)
+                        },
+                        err => {
+                            console.log(err.message);
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/checkpoint.service.js]: delete this!

File: myTrip/static/src/checkpoint/checkpoint.service.js [hunk 16]
```diff
@@ -31,9 +30,30 @@ function getAllCheckpoints(trip_id) {
...
+                            center.lat=49.832721
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/checkpoint.service.js]: move lng, lat to const

File: myTrip/static/src/checkpoint/checkpointDetails.js [hunk 17]
```diff
@@ -20,23 +29,26 @@ let images = [
 const styles = {
     gridList: {
         width: 200,
-        height: 200,
+        height: 185,
         overflowY: 'auto',
     },
 };
-
 // import {store} from './testing-page-for-checkpoints.js'
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/checkpointDetails.js]: delete this!

File: myTrip/static/src/checkpoint/checkpointDetails.js [hunk 18]
```diff
@@ -20,23 +29,26 @@ let images = [
...
       super(props);
         this.state = {
+            open: false,
             title:'',
-            description:''
+            description:'',
+            author: this.props.trip.user,
+            userId: userId
         }
       // this.updateState = this.updateState.bind(this);
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/checkpointDetails.js]: delete this!

File: myTrip/static/src/checkpoint/error-notification.js [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+        super()
+        this.state={
+            errors:{},
+            open:false
+        }
+    }
+
+    componentWillReceiveProps(nextProps) {
+        if(nextProps){
+            console.log('notificator recieved props',nextProps)
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/error-notification.js]: delete this!

File: myTrip/static/src/checkpoint/error-notification.js [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+            }
+        }
+    }
+
+    handleRequestClose = () => {
+        this.setState({open:false,});
+    };
+
+    render(){
+        console.log('changes in notificator',this.state)
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/error-notification.js]: delete this!

File: myTrip/static/src/checkpoint/error-notification.js [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+        console.log('errors-notificator component',this.state.errors)
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/error-notification.js]: delete this!

File: myTrip/static/src/checkpoint/main.less [hunk 22]
```diff
@@ -63,4 +67,4 @@
     flex-direction: row;
     justify-content: space-around;
     align-items: center;
-}
+}
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/main.less]: add endline

File: myTrip/static/src/checkpoint/map.js [hunk 23]
```diff
@@ -1,92 +1,271 @@
...
 
-    render(){
-        const mapContainer = <div style={{height: '100%', width:'100%'}}></div>
-        // if(this.props.checkpoints != null && this.props.checkpoints[0]){
-        const center = {
-            lat : 49.832721,
-            lng : 23.999003
+    componentWillReceiveProps(nextProps){
+        if (nextProps) {
+            console.log('recieved props',nextProps)
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: delete this!

File: myTrip/static/src/checkpoint/map.js [hunk 24]
```diff
@@ -1,92 +1,271 @@
...
+            let len=nextProps.checkpoints.length;
+            let lat;
+            let lng;
+            if(nextProps.active!=null){
+                // center if there is active
+                lat=nextProps.active.latitude
+                lng=nextProps.active.longitude
+                console.log(1)
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: delete this!

File: myTrip/static/src/checkpoint/map.js [hunk 25]
```diff
@@ -1,92 +1,271 @@
...
+                this.setState({ checkpoints:nextProps.checkpoints,
+                                active:nextProps.active,
+                                center:{lat:lat,lng:lng}});
+            }else{
+                if(nextProps.checkpoints.length==0){
+                    var self = this
+                    // center if no checkpoints at all getting position of user if unable hardcode it 
+                    console.log(2)
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: delete this!

File: myTrip/static/src/checkpoint/map.js [hunk 26]
```diff
@@ -1,92 +1,271 @@
...
+                    
+                    service.CurrentPosition.then(response => {
+                        console.log('position from service',response);
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: delete this!

File: myTrip/static/src/checkpoint/map.js [hunk 27]
```diff
@@ -1,92 +1,271 @@
...
+                        this.setState({ checkpoints:nextProps.checkpoints,
+                                active:nextProps.active,
+                                center:response});
+                    },error => {
+                        console.log('error from promise',error);
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: delete this!

File: myTrip/static/src/checkpoint/map.js [hunk 28]
```diff
@@ -1,92 +1,271 @@
...
+                    var self = this
+                    // center if no checkpoints at all getting position of user if unable hardcode it 
+                    console.log(2)
+                    
+                    service.CurrentPosition.then(response => {
+                        console.log('position from service',response);
+                        this.setState({ checkpoints:nextProps.checkpoints,
+                                active:nextProps.active,
+                                center:response});
+                    },error => {
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: fix tab/space

File: myTrip/static/src/checkpoint/map.js [hunk 29]
```diff
@@ -1,92 +1,271 @@
...
+                        this.setState({ checkpoints:nextProps.checkpoints,
+                                active:nextProps.active,
+                                center:error});
+                    })
+                    
+
+                }else{
+                    if(this.state.map){
+                        // center from loaded map (adding and deleteng without centring)
+                        console.log(4,this.state.map.props.center.lat)
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: delete this!

File: myTrip/static/src/checkpoint/map.js [hunk 30]
```diff
@@ -1,92 +1,271 @@
...
+                        lat=this.state.map.props.center.lat
+                        lng=this.state.map.props.center.lng
+                        this.setState({ checkpoints:nextProps.checkpoints,
+                                active:nextProps.active,
+                                center:{lat:lat,lng:lng}});
+                    }else{
+                        // /center if there are checkpoints gets latest position
+                        lat=nextProps.checkpoints[len-1].latitude
+                        lng=nextProps.checkpoints[len-1].longitude
+                        console.log(3)
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: delete this!

File: myTrip/static/src/checkpoint/map.js [hunk 31]
```diff
@@ -1,92 +1,271 @@
...
+                                center:response});
+                    },error => {
+                        console.log('error from promise',error);
+                        this.setState({ checkpoints:nextProps.checkpoints,
+                                active:nextProps.active,
+                                center:error});
+                    })
+                    
+
+                }else{
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js:54]: fix space

File: myTrip/static/src/checkpoint/map.js [hunk 32]
```diff
@@ -1,92 +1,271 @@
...
+                    if(this.state.map){
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: fix space

File: myTrip/static/src/checkpoint/map.js [hunk 33]
```diff
@@ -1,92 +1,271 @@
...
+                        // center from loaded map (adding and deleteng without centring)
+                        console.log(4,this.state.map.props.center.lat)
+                        lat=this.state.map.props.center.lat
+                        lng=this.state.map.props.center.lng
+                        this.setState({ checkpoints:nextProps.checkpoints,
+                                active:nextProps.active,
+                                center:{lat:lat,lng:lng}});
+                    }else{
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js:62]: fix space

File: myTrip/static/src/checkpoint/map.js [hunk 34]
```diff
@@ -1,92 +1,271 @@
...
+                }
+            } 
         }
+    }
+
+    mapLoaded=(map)=>{
+        if(this.state.ma!=null)
+            return
+        this.setState({map:map})
+        console.log('map loaded');
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: delete this!

File: myTrip/static/src/checkpoint/map.js [hunk 35]
```diff
@@ -1,92 +1,271 @@
...
+                        this.setState({ checkpoints:nextProps.checkpoints,
+                                active:nextProps.active,
+                                center:{lat:lat,lng:lng}});
+                    }
+                }
+            } 
         }
+    }
+
+    mapLoaded=(map)=>{
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: fix tab/space  ` mapLoaded = map => {`

File: myTrip/static/src/checkpoint/map.js [hunk 36]
```diff
@@ -1,92 +1,271 @@
...
-            let list = this.props.checkpoints
-            if (this.props.active != null) {
-                center.lat = this.props.active.latitude,
-                center.lng = this.props.active.longitude
-            } else {
-                let len = this.props.checkpoints.length;
-                center.lat = list[len-1].latitude
-                center.lng = list[len-1].longitude
-                
+    handleMapClick=(map)=>{
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: `=  map => {`

File: myTrip/static/src/checkpoint/map.js [hunk 37]
```diff
@@ -1,92 +1,271 @@
...
+        // if(this.state.active || userId() != this.props.trip.user)
+        if(this.state.active || userId() != this.props.trip.user){
+            return
+        }else{
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: fix tab/space

File: myTrip/static/src/checkpoint/map.js [hunk 38]
```diff
@@ -1,92 +1,271 @@
...
+            const longitude=map.latLng.lng();
+            const latitude=map.latLng.lat();
+            const description='';
+            let position_number;
+            if(this.state.checkpoints == undefined || this.state.checkpoints.length == 0){
+                position_number=1;
+            }else{
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: fix tab/space

File: myTrip/static/src/checkpoint/map.js [hunk 39]
```diff
@@ -1,92 +1,271 @@
...
+                            onRightclick={()=>this.handleRightClick(point)}
+                            icon={'http://maps.google.com/mapfiles/ms/icons/yellow-dot.png'}>
+                                {infoWindow}
+                        </Marker>
+        });
+        return markers;
+    }
+
+    handleZoom = ()=> {
+        console.log(this.state.map.getZoom());
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: delete this!

File: myTrip/static/src/checkpoint/map.js [hunk 40]
```diff
@@ -1,92 +1,271 @@
...
+    }
+
+    handleDragMap = () => {
+        console.log('map moved', this.state.map.getCenter().lat(),this.state.map.getCenter().lng())
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: delete this!

File: myTrip/static/src/checkpoint/map.js [hunk 41]
```diff
@@ -1,92 +1,271 @@
...
+        this.setState({center:{lat:this.state.map.getCenter().lat(),lng:this.state.map.getCenter().lng()}})
+    }
+
+    handleMarkerDrag = marker => {
+        console.log('moved',marker.latLng.lat(),marker.latLng.lng())
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: delete this!

File: myTrip/static/src/checkpoint/map.js [hunk 42]
```diff
@@ -1,92 +1,271 @@
...
+        console.log(marker)
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: delete this!

File: myTrip/static/src/checkpoint/map.js [hunk 43]
```diff
@@ -1,92 +1,271 @@
...
+            description,
+            position_number,
+            source_url,
+            trip_id,
+            checkpoint_id)
+    }
+
+    render(){
+        const mapContainer=<div style={{height:'100%', width:'100%'}}></div>
+        console.log('state changed',this.state)
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/map.js]: delete this!

File: myTrip/static/src/checkpoint/reducers/checkpoint-list-reducer.js [hunk 44]
```diff
@@ -7,6 +7,8 @@ export default function (state = null, action) {
             return action.payload;
 
         case 'CREATE-CHECKPOINT-UPDATE_LIST':
+            console.log('reducer createcheckpoint',action.payload)
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/reducers/checkpoint-list-reducer.js]: delete this!

File: myTrip/static/src/checkpoint/reducers/checkpoint-list-reducer.js [hunk 45]
```diff
@@ -7,6 +7,8 @@ export default function (state = null, action) {
...
+            console.log('reducer createcheckpoint',state)
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/reducers/checkpoint-list-reducer.js]: delete this!

File: myTrip/static/src/checkpoint/reducers/error-reducer.js [hunk 46]
```diff
@@ -0,0 +1,9 @@
+export default function (state = null, action) {
+    switch (action.type) {
+
+        case 'ERROR':
+            console.log('error reducer payload',action.payload)
```
Comment (lhalam) [inline myTrip/static/src/checkpoint/reducers/error-reducer.js]: delete this!

**Review by black-art — APPROVED** [review]

**Pull request merged**